cmip6_preprocessing/preprocessing.py

```diff
@@ -202,30 +209,24 @@ def maybe_fix_non_unique(data, pad=False):
     return ds
 
 
-def unit_conversion_dict():
-    """Units conversion database"""
-    unit_dict = {"m": {"centimeters": 1 / 100}}
-    return unit_dict
-
-
-def correct_units(ds, verbose=False, stric=False):
-    "Converts coordinates into SI units using `unit_conversion_dict`"
-    unit_dict = unit_conversion_dict()
-    ds = ds.copy()
-    # coordinate conversions
-    for co, expected_unit in [("lev", "m")]:
-        if co in ds.coords:
-            if "units" in ds.coords[co].attrs.keys():
-                unit = ds.coords[co].attrs["units"]
-                if unit != expected_unit:
-                    if unit in unit_dict[expected_unit].keys():
-                        factor = unit_dict[expected_unit][unit]
-                        ds.coords[co] = ds.coords[co] * factor
-                        ds.coords[co].attrs["units"] = expected_unit
-                    else:
-                        warnings.warn("No conversion found in unit_dict")
-            else:
-                warnings.warn(f'{ds.attrs["source_id"]}: No units found for {co}')
+def correct_units(ds):
+    "Converts coordinates into SI units using pint-xarray"
+    # codify units with pint
+    # Perhaps this should be kept separately from the fixing?
+    # See https://github.com/jbusecke/cmip6_preprocessing/pull/160#discussion_r667041858
+
+    # exclude salinity from the quantification (see https://github.com/jbusecke/cmip6_preprocessing/pull/160#issuecomment-878627027 for details)
+    overrides = {name: None for name, var in ds.variables.items() if name in ["so"]}
+    quantified = ds.pint.quantify(overrides)
+
+    desired_units = _desired_units()
+    for var, target_unit in desired_units.items():
+        if var in quantified:
+            if "units" in quantified[var].attrs.keys():
+                # do we need to check this, or is pint smart enough to not touch the units, if its already the one we want?
+                # if ds["lev"].units != "m":
+                quantified = quantified.pint.to({var: target_unit})
```

Comment:
`Dataset.pint.to` is capable of converting multiple variables at once. I would construct a filtered "target_units" dict before passing the whole dictionary at once. This would also give a nicer error behavior in case multiple conversions fail: you get all failing variables at once instead of having to fix one only to notice that another also failed.


Also the `"units" in quantified[var].attrs` check will only evaluate to true for dimension coordinates, all other variables either don't have units or their `"units"` attribute has been removed after quantifying (of course, if you wanted to only convert dimension coordinates feel free to ignore... but then it might be more readable to explicitly check that).

```suggestion
    target_units = {
        var: target_units
        for var, target_unit in desired_units.items()
        if var in quantified
    }
    converted = quantified.pint.to(target_units)
```

Regarding the comment that would be removed: yes, `pint` returns immediately if it notices that there's nothing to do (see [`pint.registry.BaseRegistry.convert`](https://github.com/hgrecco/pint/blob/abe384033bffa7627387eeaec297f0a75777b65d/pint/registry.py#L938-L939))